lib/ci/queue/redis/worker.rb

```diff
@@ -28,18 +33,22 @@ def master?
 
         def poll
           wait_for_master
-          while test = reserve
-            yield test
-            acknowledge(test)
+          until shutdown_required? || empty?
+            if test = reserve
+              yield test
+            else
+              sleep 0.05
+            end
```

Comment:
Nevermind, I think I get this now. The queue may be empty, but we have to wait for the retry queue to also be empty.